File: js/Projectile.js
class Projectile extends Phaser.Physics.Arcade.Sprite {
    constructor(scene, x, y, direction, owner, projectileType, chargeLevel = 1) {
        super(scene, x, y, 'projectile');
        
        this.scene = scene;
        this.direction = direction;
        this.owner = owner;
        this.type = projectileType;
        this.active = true;
        this.lifetime = 100;
        this.chargeLevel = chargeLevel;
        
        // Initialize based on type
        this.initProjectileType();
        
        // Add to scene
        scene.add.existing(this);
        scene.physics.add.existing(this);
        
        // Set physics properties
        this.setCollideWorldBounds(false);
        this.body.setAllowGravity(false);
        
        // Create visual representation
        this.createVisual();
    }
    
    initProjectileType() {
        switch(this.type) {
            case "fireball":
                this.radius = 15;
                this.speed = 10;
                this.damage = 12;
                break;
            case "ice_shard":
                this.radius = 12;
                this.speed = 14;
                this.damage = 7;
                this.vy = 0;
                this.body.setAllowGravity(true);
                this.body.setGravityY(300);
                break;
            case "missile":
                this.radius = 8;
                this.speed = 16;
                this.damage = 6;
                this.lifetime = 40;
                break;
            case "arcane_orb":
                this.radius = 20;
                this.speed = 4;
                this.damage = 16;
                this.lifetime = 150;
                break;
            case "charge_shot":
                this.radius = 10 + Math.floor(this.chargeLevel / 10);
                this.speed = 12 + Math.floor(this.chargeLevel / 20);
                this.damage = 8 + Math.floor(this.chargeLevel / 8);
                break;
            case "fire_breath":
                this.radius = 25;
                this.speed = 3;
                this.damage = 14;
                this.lifetime = 25;
                break;
        }
        
        this.setSize(this.radius * 2, this.radius * 2);
        this.setVelocityX(this.speed * this.direction);
    }
    
    createVisual() {
        // Create graphics for the projectile
        this.graphics = this.scene.add.graphics();
        this.graphics.setDepth(5); // Draw projectiles above background, below characters
        this.updateVisual();
    }
    
    updateVisual() {
        this.graphics.clear();
        this.graphics.setPosition(0, 0); // Graphics at world origin
        
        // Set position - all coordinates are world coordinates
        const drawX = this.x;
        const drawY = this.y;
        
        switch(this.type) {
            case "fireball":
                this.graphics.fillStyle(COLORS.ORANGE);
                this.graphics.fillCircle(drawX, drawY, this.radius);
                this.graphics.fillStyle(COLORS.YELLOW);
                this.graphics.fillCircle(drawX, drawY, this.radius - 5);
                this.graphics.fillStyle(COLORS.RED);
                this.graphics.fillCircle(drawX, drawY, this.radius - 10);
                break;
            case "ice_shard":
                // Diamond shape - 4 points (diamond/polygon)
                this.graphics.fillStyle(COLORS.CYAN);
                this.graphics.fillTriangle(
                    drawX, drawY - this.radius,
                    drawX + this.radius/2, drawY,
                    drawX, drawY + this.radius
                );
                this.graphics.fillTriangle(
                    drawX, drawY - this.radius,
                    drawX - this.radius/2, drawY,
                    drawX, drawY + this.radius
                );
                this.graphics.lineStyle(2, COLORS.WHITE);
                this.graphics.strokeTriangle(
                    drawX, drawY - this.radius,
                    drawX + this.radius/2, drawY,
                    drawX, drawY + this.radius
                );
                this.graphics.strokeTriangle(
                    drawX, drawY - this.radius,
                    drawX - this.radius/2, drawY,
                    drawX, drawY + this.radius
                );
                break;
            case "missile":
                this.graphics.fillStyle(COLORS.YELLOW);
                this.graphics.fillCircle(drawX, drawY, this.radius);
                this.graphics.fillStyle(COLORS.ORANGE);
                this.graphics.fillRect(drawX - this.radius, drawY - this.radius/2, this.radius*2, this.radius);
                break;
            case "arcane_orb":
                this.graphics.fillStyle(COLORS.PURPLE);
                this.graphics.fillCircle(drawX, drawY, this.radius);
                this.graphics.fillStyle(COLORS.PINK);
                this.graphics.fillCircle(drawX, drawY, this.radius - 4);
                this.graphics.fillStyle(COLORS.WHITE);
                this.graphics.fillCircle(drawX, drawY, this.radius - 12);
                break;
            case "charge_shot":
                this.graphics.fillStyle(COLORS.GREEN);
                this.graphics.fillCircle(drawX, drawY, this.radius);
                this.graphics.fillStyle(COLORS.YELLOW);
                this.graphics.fillCircle(drawX, drawY, this.radius * 0.7);
                this.graphics.fillStyle(COLORS.WHITE);
                this.graphics.fillCircle(drawX, drawY, this.radius * 0.3);
                break;
            case "fire_breath":
                // Draw cloud of fire particles with alpha - use main graphics object
                this.graphics.fillStyle(Phaser.Display.Color.GetColor(255, 100, 0), 0.6);
                this.graphics.fillCircle(drawX, drawY, this.radius);
                this.graphics.fillStyle(Phaser.Display.Color.GetColor(255, 255, 0), 0.4);
                this.graphics.fillCircle(drawX + 5, drawY, this.radius/2);
                this.graphics.fillCircle(drawX - 5, drawY, this.radius/2);
                break;
        }
    }
    
    update() {
        if (this.type === "fire_breath") {
            this.radius = Math.max(5, this.radius - 1);
            this.setSize(this.radius * 2, this.radius * 2);
        }
        
        this.lifetime -= 1;
        this.updateVisual();
        
        // Check if should be destroyed
        if (this.lifetime <= 0 || 
            this.x < -50 || 
            this.x > GAME_CONFIG.WIDTH + 50 || 
            this.y > GAME_CONFIG.HEIGHT + 50) {
            this.destroy();
        }
    }
    
    destroy() {
        this.active = false;
        if (this.graphics) {
            this.graphics.destroy();
        }
        super.destroy();
    }
}
// <-- The extra '}' was removed from here

File: gamePython/projectile.py
import pygame
import math
from constants import *

class Projectile:
    def __init__(self, x, y, direction, owner, projectile_type, **kwargs):
        self.x = x
        self.y = y
        self.direction = direction
        self.owner = owner
        self.type = projectile_type
        self.active = True
        self.lifetime = 100
        
        if projectile_type == "fireball":
            self.radius = 15
            self.speed = 10
            self.damage = 12
        elif projectile_type == "ice_shard":
            self.radius = 12
            self.speed = 14
            self.damage = 7
            self.vy = 0
        elif projectile_type == "missile":
            self.radius = 8
            self.speed = 16
            self.damage = 6
            self.lifetime = 40
        elif projectile_type == "arcane_orb":
            self.radius = 20
            self.speed = 4
            self.damage = 16
            self.lifetime = 150
        elif projectile_type == "charge_shot":
            charge_level = kwargs.get("charge_level", 1)
            self.radius = 10 + int(charge_level / 10)
            self.speed = 12 + int(charge_level / 20)
            self.damage = 8 + int(charge_level / 8)
        elif projectile_type == "fire_breath":
            self.radius = 25
            self.speed = 3
            self.damage = 14
            self.lifetime = 25 # Short range
    
    def update(self):
        self.x += self.speed * self.direction
        if self.type == "ice_shard":
            self.vy += 0.3
            self.y += self.vy
        
        if self.type == "fire_breath":
            self.radius = max(5, self.radius - 1) # Shrinks
        
        self.lifetime -= 1
        if self.lifetime <= 0 or self.x < -50 or self.x > WIDTH + 50 or self.y > HEIGHT + 50:
            self.active = False
    
    def draw(self, screen):
        if self.type == "fireball":
            pygame.draw.circle(screen, ORANGE, (int(self.x), int(self.y)), self.radius)
            pygame.draw.circle(screen, YELLOW, (int(self.x), int(self.y)), self.radius - 5)
            pygame.draw.circle(screen, RED, (int(self.x), int(self.y)), self.radius - 10)
        elif self.type == "ice_shard":
            points = [(self.x, self.y - self.radius), (self.x + self.radius/2, self.y),
                      (self.x, self.y + self.radius), (self.x - self.radius/2, self.y)]
            pygame.draw.polygon(screen, CYAN, points)
            pygame.draw.polygon(screen, WHITE, points, 2)
        elif self.type == "missile":
            pygame.draw.circle(screen, YELLOW, (int(self.x), int(self.y)), self.radius)
            pygame.draw.rect(screen, ORANGE, (self.x - self.radius, self.y - self.radius/2, self.radius*2, self.radius))
        elif self.type == "arcane_orb":
            pygame.draw.circle(screen, PURPLE, (int(self.x), int(self.y)), self.radius)
            pygame.draw.circle(screen, PINK, (int(self.x), int(self.y)), self.radius - 4)
            pygame.draw.circle(screen, WHITE, (int(self.x), int(self.y)), self.radius - 12)
        elif self.type == "charge_shot":
            pygame.draw.circle(screen, GREEN, (int(self.x), int(self.y)), self.radius)
            pygame.draw.circle(screen, YELLOW, (int(self.x), int(self.y)), int(self.radius * 0.7))
            pygame.draw.circle(screen, WHITE, (int(self.x), int(self.y)), int(self.radius * 0.3))
        elif self.type == "fire_breath":
            # Draw a cloud of fire particles
            color1 = (255, 100, 0, 150) # Orange
            color2 = (255, 255, 0, 100) # Yellow
            
            s = pygame.Surface((self.radius*2, self.radius*2), pygame.SRCALPHA)
            pygame.draw.circle(s, color1, (self.radius, self.radius), self.radius)
            pygame.draw.circle(s, color2, (self.radius+5, self.radius), int(self.radius/2))
            pygame.draw.circle(s, color2, (self.radius-5, self.radius), int(self.radius/2))
            screen.blit(s, (self.x - self.radius, self.y - self.radius))
    
    def get_rect(self):
        return pygame.Rect(self.x - self.radius, self.y - self.radius, 
                          self.radius * 2, self.radius * 2)